stm32-modules/heater-shaker/firmware/heater_task/heater_hardware.c

```diff
@@ -132,15 +191,36 @@ void heater_hardware_release_pg_latch() {
                       GPIO_PIN_SET);
 }
 
+void heater_hardware_power_disable(heater_hardware* hardware) {
+    hw_internal* internal = (hw_internal*)hardware->hardware_internal;
```

Comment:
this is kind of the "private data" idiom in c, where you have a struct definition in your header (heater_hardware) with stuff you expect outside code  to initialize for you - like the conversion_complete callback - and a `void*`,  which erases types, to something you define only in your c file  and thus that only you know about, which is `hw_internal`. The downside is you have to make a local copy with a cast like this every time because the compiler needs to know what's on the other side of the void*.